### PR TITLE
KeyNav 2.0

### PR DESCRIPTION
**Pull request by brboi**

### CODE REVIEW

File: addons/wowl/static/src/services/hotkey.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+              callback: listener[1].callback,
+              hotkey: listenerHotkey,
+            });
+          }
+        }
+        return listenersToNotify;
+      }, []);
+
+      for (const listener of listenersToNotify) {
+        listener.callback(listener.hotkey);
```
Comment (brboi) [inline addons/wowl/static/src/services/hotkey.js]: multiple candidates listeners ??

File: addons/wowl/doc/services/hotkey.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
+# Hotkey Service
+
+| Technical name | Dependencies                           |
+| -------------- | -------------------------------------- |
+| `hotkey`       | `dialog_manager`, `localization`, `ui` |
```
Comment (ged-odoo) [inline addons/wowl/doc/services/hotkey.md]: dialog

File: addons/wowl/doc/services/hotkey.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+## API
+
+The `hotkey` service provides the following API:
+
+- `subscribe(sub: HotkeySubscription): number`:
+  it ask the service to call the given callback when a matching hotkey is pressed.
+  The service returns a token you can use to unsubscribe later on.
+  An `HotkeySubscription` object should take the form of:
+  `{ hotkeys: string[], callback: (hotkey:string)=>void, hint?: string }`.
```
Comment (ged-odoo) [inline addons/wowl/doc/services/hotkey.md]: `subscribe(hotkey: string, callback)`

Comment (ged-odoo) [reply addons/wowl/doc/services/hotkey.md]: also, no documentation for return value

Comment (ged-odoo) [reply addons/wowl/doc/services/hotkey.md]:
```ts
interface HKSub {
  hotkey: string;
  callback: ...;
  hint?: string
}
```

File: addons/wowl/doc/services/hotkey.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  When the defined hotkey is pressed, the element gets clicked.
+
+## Examples
+
+### `useHotkeys` hook
+```js
+class MyComponent extends Component {
+  setup() {
+    useHotkeys([
+      { hotkeys: ["a", "shift-a"], callback: this.callback },
```
Comment (ged-odoo) [inline addons/wowl/doc/services/hotkey.md]:
```js

    { hotkey: "a", callback: () => this.activateThing() },
```

File: addons/wowl/doc/services/hotkey.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+- `[data-hotkey]`: an HTML attribute taking an hotkey definition.
+  When the defined hotkey is pressed, the element gets clicked.
+
+## Examples
+
+### `useHotkeys` hook
+```js
+class MyComponent extends Component {
+  setup() {
+    useHotkeys([
```
Comment (ged-odoo) [inline addons/wowl/doc/services/hotkey.md]: `useHotkey`

File: addons/wowl/static/src/services/ui_service.js [hunk 6]
```diff
@@ -1,21 +1,72 @@
...
+ * This hook will set the UI ownership
+ * when the caller component will mount/unmount.
+ * The caller component could pass a `t-ref` value of its template
+ * to delegate the UI ownership to another element than itself.
+ *
+ * @param {string?} refName
+ */
+export function useUIOwnership(refName) {
+  const uiService = useService("ui");
+  const owner = (refName && hooks.useRef(refName)) || Component.current;
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/ui_service.js]:
```js
  const owner = refName ? hooks.useRef(refName) : Component.current;

```

Comment (ged-odoo) [reply addons/wowl/static/src/services/ui_service.js]: let el = null;

Comment (ged-odoo) [reply addons/wowl/static/src/services/ui_service.js]:
```js
  let el = null;

  onMounted(checkOwnerShip);
  onPatched(checkOwnerShip);
  onWillUnmount(checkOwnerShip);
 function checkOwnerShip() {
   if (el !== owner.el) {
     el = owner.el;
     if (el) {
      takeOwnerShip(el);
    } else {
     releaseOwnership()
   }
   }
 }

```nope

Comment (ged-odoo) [reply addons/wowl/static/src/services/ui_service.js]: test with owner in t-if..

no

rather => only support cases where el is fixed (not dynamic)

File: addons/wowl/static/src/services/hotkey_service.js [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+    /**
+     * Handler for keydown events.
+     *
+     * @param {KeyboardEvent} ev
+     */
+    async function onKeydown(ev) {
+      const hotkey = getActiveHotkey(ev);
+      const infos = { hotkey, _originalEvent: ev };
+      if (await canDispatch(infos)) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/hotkey_service.js]: no async

File: addons/wowl/static/src/services/hotkey_service.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+const MODIFIERS = ["control", "shift"];
+const ALPHANUM_KEYS = "abcdefghijklmnopqrstuvwxyz0123456789".split("");
+const NAV_KEYS = [
+  "arrowleft", "arrowright", "arrowup", "arrowdown",
+  "pageup", "pagedown", "home", "end",
+  "backspace", "enter", "escape",
+]
+const AUTHORIZED_KEYS = [...ALPHANUM_KEYS, ...NAV_KEYS];
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/hotkey_service.js]: make it a set

File: addons/wowl/static/src/services/hotkey_service.js [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+     * clicks on all elements having a data-hotkey attribute matching the hotkey.
+     *
+     * @param {string} hotkey
+     */
+    function dispatch(infos) {
+      const { hotkey, _originalEvent: event } = infos;
+      const uiOwnerElement = env.services.ui.getOwner();
+
+      // 1. Dispatch actual hotkey to all matching subscriptions
+      const subs = Object.values(subscriptions)
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/hotkey_service.js]:
```js
for (let sub in subscriptions) {
  if (sub.contextOwner === uiOwnerElement YY sub.hotkey === hotkey) {
    sub.callback(hotkey);
  }
}
```

File: addons/wowl/static/src/components/dropdown/dropdown.js [hunk 10]
```diff
@@ -1,17 +1,34 @@
 /** @odoo-module **/
 import { useBus } from "../../core/hooks";
+import { useService } from "../../core/hooks";
 import { ParentClosingMode } from "./dropdown_item";
 
 const { Component, core, hooks, useState } = owl;
 
 export class Dropdown extends Component {
   constructor() {
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/dropdown/dropdown.js]: setup

File: addons/wowl/static/src/webclient/navbar/navbar.xml [hunk 11]
```diff
@@ -29,7 +29,7 @@
     </nav>
   </header>
 
-  <Dropdown t-name="wowl.NavBar.AppsMenu" class="o_navbar_apps_menu" owl="1">
+  <Dropdown t-name="wowl.NavBar.AppsMenu" togglerHotKey="'h'" class="o_navbar_apps_menu" owl="1">
```
Comment (ged-odoo) [inline addons/wowl/static/src/webclient/navbar/navbar.xml]: hotkey

File: addons/wowl/static/src/webclient/navbar/navbar.xml [hunk 12]
```diff
@@ -45,23 +45,39 @@
       t-foreach="sections"
       t-as="section"
       t-key="section.id">
-      <DropdownItem t-if="!section.childrenTree.length" payload="section">
+
+      <t t-set="sectionsVisibleCount" t-value="(sections.length - currentAppSectionsExtra.length)" />
```
Comment (ged-odoo) [inline addons/wowl/static/src/webclient/navbar/navbar.xml:49]: maybe move this in JS

**Review by ged-odoo — APPROVED** [review]

**Review by ged-odoo — APPROVED** [review]

**Pull request merged**